src/main/scala/is/hail/methods/LinearMixedRegression.scala

```diff
@@ -246,35 +251,75 @@ object DiagLMM {
-    else if (approxLogDelta == logmax)
-      fatal(s"lmmreg: failed to fit delta: ${if (useML) "ML" else "REML"} realized at delta upper search boundary e^$logmax = ${FastMath.exp(logmax)}, indicating negligible genetic component of variance. Standard linear regression may be more appropriate.")
+    if (approxLogDelta == minLogDelta)
+      fatal(s"lmmreg: failed to fit delta: ${if (useML) "ML" else "REML"} realized at delta lower search boundary e^$minLogDelta = ${FastMath.exp(minLogDelta)}, indicating negligible enviromental component of variance. The model is likely ill-specified.")
+    else if (approxLogDelta == maxLogDelta)
+      fatal(s"lmmreg: failed to fit delta: ${if (useML) "ML" else "REML"} realized at delta upper search boundary e^$maxLogDelta = ${FastMath.exp(maxLogDelta)}, indicating negligible genetic component of variance. Standard linear regression may be more appropriate.")
 
-    val searchInterval = new SearchInterval(-10d, 10d, approxLogDelta)
+    val searchInterval = new SearchInterval(minLogDelta, maxLogDelta, approxLogDelta)
     val goal = GoalType.MAXIMIZE
     val objectiveFunction = new UnivariateObjectiveFunction(if (useML) LogLkhdML else LogLkhdREML)
     val brentOptimizer = new BrentOptimizer(5e-8, 5e-7) // tol = 5e-8 * abs((ln(delta))) + 5e-7 <= 1e-6
     val logDeltaPointValuePair = brentOptimizer.optimize(objectiveFunction, goal, searchInterval, MaxEval.unlimited)
 
-    val logDelta = logDeltaPointValuePair.getPoint
+    val maxlogDelta = logDeltaPointValuePair.getPoint
     val maxLogLkhd = logDeltaPointValuePair.getValue
 
-    if (math.abs(logDelta - approxLogDelta) > 1.0 / pointsPerUnit) {
-      warn(s"lmmreg: the difference between the optimal value $approxLogDelta of ln(delta) on the grid and the optimal value $logDelta of ln(delta) by Brent's method exceeds the grid resolution of ${1.0 / pointsPerUnit}. Plot the values over the full grid to investigate.")
+    if (math.abs(maxlogDelta - approxLogDelta) > 1d / pointsPerUnit) {
+      warn(s"lmmreg: the difference between the optimal value $approxLogDelta of ln(delta) on the grid and" +
+        s"the optimal value $maxlogDelta of ln(delta) by Brent's method exceeds the grid resolution" +
+        s"of ${1d / pointsPerUnit}. Plot the values over the full grid to investigate.")
     }
 
-    (FastMath.exp(logDelta), Some(maxLogLkhd), Some(gridLogLkhd))
+    val epsilon = 1d / pointsPerUnit
+
+    // three values of ln(delta) right of, at, and left of the MLE
+    val z1 = maxlogDelta + epsilon
+    val z2 = maxlogDelta
+    val z3 = maxlogDelta - epsilon
+
+    // three values of h2 = sigmoid(-ln(delta)) left of, at, and right of the MLE
+    val x1 = sigmoid(-z1)
+    val x2 = sigmoid(-z2)
+    val x3 = sigmoid(-z3)
+
+    // corresponding values of logLkhd
+    val y1 = if (useML) LogLkhdML.value(z1) else LogLkhdREML.value(z1)
```

Comment:
Don't worry about it since I think I already changed it in my Low-rank branch, but you have a lot of things like `if (useML) ..... LogLkhdML ..... else ......LogLkhdREML ......`. They could all be replaced by just doing something at the beginning of the function like `val liklihoodFunction = if (useML) LogLkhdML else LogLkhdREML`. Just something to think about as a way to avoid if statements everywhere.